apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FieldPath.cs

```diff
@@ -0,0 +1,157 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Linq;
+using System.Text;
+
+namespace Google.Cloud.Firestore.Data
+{
+    /// <summary>
+    /// An immutable path of field names, used to identify parts of a document.
+    /// </summary>
+    public sealed class FieldPath : IEquatable<FieldPath>
+    {
+        private static readonly char[] s_dotSplit = { '.' };
+
+        /// <summary>
+        /// Regular expression to verify that dot-separated field paths do not contain ~*[]/ characters.
+        /// </summary>
+        private static readonly char[] s_prohibitedCharacters = { '~', '*', '[', ']', '/' };
+
+        /// <summary>
+        /// Sentinel field path to refer to the ID of a document. Used in queries to sort or filter
+        /// by the document ID.
+        /// </summary>
+        public static FieldPath DocumentId { get; } = new FieldPath("__name__");
+
+        /// <summary>
+        /// An empty field path, appropriate for appending
+        /// </summary>
+        internal static FieldPath Empty { get; } = new FieldPath(new string[0], true);
+
+        /// <summary>
+        /// Segments of the path: do not mutate or expose publicly!
+        /// </summary>
+        internal string[] Segments { get; }
+
+        private string _encodedPath;
+
+        internal string EncodedPath => _encodedPath ?? (_encodedPath = GetCanonicalPath(Segments));
+
+        /// <summary>
+        /// Constructor that trusts the value
+        /// </summary>
+        /// <param name="segments">The segments</param>
+        /// <param name="ignored">Ignored</param>
+        private FieldPath(string[] segments, bool ignored)
+        {
+            Segments = segments;
+        }
+
+        /// <summary>
+        /// Creates a path from multiple segments. Each segment is treated verbatim: it may contain dots, which will lead to the segment being escaped in the
+        /// path's string representation.
+        /// </summary>
+        /// <param name="segments">The segments of the path. This must not be null or empty, and it must not contain any null or empty elements.</param>
+        public FieldPath(params string[] segments)
+        {
+            Segments = GaxPreconditions.CheckNotNull(segments, nameof(segments)).ToArray();
+            GaxPreconditions.CheckArgument(segments.Length != 0, nameof(segments), "Path must not be empty");
+            GaxPreconditions.CheckArgument(Segments.All(n => !string.IsNullOrEmpty(n)), nameof(segments), "Path must not contain null or empty names");
+        }
+
+        internal static FieldPath FromDotSeparatedString(string field)
```

Comment:
Why doesn't this support escaped dot-separated-string paths?
Or, being internal, is it simply not required?